File: observer/WeatherStation.cpp
//observer c++
#include <iostream>
#include <list>
using namespace std;

//Observer
class Observer{
public:
	virtual void update(float temp,float humidity,float pressure) = 0;
};

//DisplayElement
class DisplayElement{
public:
	virtual void display() = 0;
};

//Subject
class Subject{
public:
	virtual void registerObserver(Observer* o) = 0;
	virtual void removeObserver  (Observer* o) = 0;
	virtual void notifyObservers ()            = 0;
	
};

class WeatherData : public Subject{
public:
	WeatherData(){}
	void registerObserver(Observer* o){
		if(o)
			observers.push_back(o);
	}
	void removeObserver(Observer* o){
		auto it = observers.begin();                   //begin()返回的类型由对象是否为const决定
		while(it != observers.end()){                  //只要执行了erase(it),那么it就会变得无效，
													   //那么执行it++就肯定会出错,erase 返回指向删除后位置，要写 it = erase(it);
			if(*it == o)
				it = observers.erase(it);
		}
	}
	void notifyObservers(){
		for(Observer* obs : observers){
			obs->update(temperature,humidity,pressure);
		}
	}
	void measurementsChanged(){
		notifyObservers();
	}
	void setMeasurements(float temperature,float humidity,float pressure){
		this->temperature = temperature;
		this->humidity    = humidity;
		this->pressure    = pressure;
		measurementsChanged();
	}

private:
	list<Observer*> observers;
	float temperature;
	float humidity;
	float pressure;
};

//Current Display
class CurrentConditonDisplay : public Observer,DisplayElement{

public:
	//抽象类型不能用作形参类型，函数返回类型
	CurrentConditonDisplay(WeatherData* weatherData){
		this->weatherData = weatherData;
		weatherData->registerObserver(this);
	}
	void update(float temperature,float humidity,float pressure){
		this->temperature = temperature;
		this->humidity    = humidity;
		display();
	}

	void display(){
		cout<<"\r\nCurrent condition:"<<temperature<<"F degree and "<<humidity<<"%humidity";
	}
	
private:
	float temperature;
	float humidity;
	WeatherData* weatherData;
};

//
int main(){
	WeatherData* weatherData = new WeatherData();        //WeatherData
	CurrentConditonDisplay* currentDisplay = new CurrentConditonDisplay(weatherData);

	weatherData->setMeasurements(80,65,26.3f);
	weatherData->setMeasurements(89,61,30.3f);
	
	//remove
	weatherData->removeObserver(currentDisplay);
	weatherData->setMeasurements(76,67,12.3f);

	return 0;
}

File: observer/WeatherStation.java
import java.util.ArrayList;

//Subject
interface Subject{
	public void registerObserver(Observer o);
	public void removeObserver(Observer o);
	public void notifyObservers();
}

class WeatherData implements Subject{
	private ArrayList<Observer> observers;
	private float temperature;
	private float humidity;
	private float pressure;
	
	public WeatherData(){
		observers = new ArrayList<Observer>();
	}
	//+
	public void registerObserver(Observer o){
		observers.add(o);
	}
	//-
	public void removeObserver(Observer o){
		int i = observers.indexOf(o);
		if(i >= 0)
			observers.remove(i);
	}
	// )))
	public void notifyObservers(){
		for (Observer observer : observers){
			observer.update(temperature,humidity,pressure);
		}
	}
	public void measurementsChanged(){
		notifyObservers();
	}
	public void setMeasurements(float temperature,float humidity,float pressure){
		this.temperature = temperature;
		this.humidity    = humidity;
		this.pressure    = pressure;
		measurementsChanged();
	}
}
//Observer
interface Observer{
	public void update(float temp,float humidity,float pressure);
}

interface DisplayElement{
	public void display();
}

//Current Display
class CurrentConditonDisplay implements Observer,DisplayElement{
	private float temperature;
	private float humidity;
	private Subject weatherData;                           //Subject
	
	public CurrentConditonDisplay(Subject weatherData){
		this.weatherData = weatherData;
		weatherData.registerObserver(this);
	}
	public void update(float temperature,float humidity,float pressure){
		this.temperature = temperature;
		this.humidity    = humidity;
		display();
	}

	public void display(){
		System.out.println("\r\nCurrent condition:" + temperature + "F degree and " + humidity + "%humidity");
	}
}
public class WeatherStation{
	public static void main(String[] args){
		WeatherData weatherData = new WeatherData();        //WeatherData
		CurrentConditonDisplay currentDisplay = new CurrentConditonDisplay(weatherData);
		//CurrentConditonDisplay currentDisplay2 = new CurrentConditonDisplay(weatherData);
		
		weatherData.setMeasurements(80,65,26.3f);
		weatherData.setMeasurements(89,61,30.3f);
		
		//remove
		weatherData.removeObserver(currentDisplay);
		weatherData.setMeasurements(76,67,12.3f);
	}
}